[1.4] Implement NPC Happiness!

kittenchilly opened the pull request:
Pretty self-explanatory. Adds a new hook for ModNPC and GlobalNPC called NPCHappiness that allows you to modify NPC biome and neighbor preferences. Also adds ExampleMod stuff for implementation, localization, etc.

kittenchilly committed:
NPCHappiness hook

kittenchilly committed:
Localization stuff. It works!

kittenchilly committed:
Fix something

kittenchilly committed:
Typo fix

kittenchilly committed:
ExampleNPCHappiness

Steviegt6 commented on patches/tModLoader/Terraria/ModLoader/Mod.cs:
```diff
@@ -414,7 +414,7 @@ public void AddConfig(string name, ModConfig mc)
 		/// </summary>
 		/// <param name="key">The key for the ModTranslation. The full key will be Mods.ModName.key</param>
```
> The full key will be Mods.ModName.key

From the looks of it, this isn't true anymore (and should probably be changed). This is just from a quick glance though, tell me if I'm wrong.

kittenchilly replied:
Oh oops, guess I missed that.

kittenchilly committed:
Merge branch '1.4' into 1.4_happiness

kittenchilly committed:
Fix for some new changes.

Solxanich reviewed (changes requested):
Regarding the localization change:
I would recommend undoing it as, although straight-forward in what you did, it has potentially far cascading consequences that we won't want to check through all.

Instead, I would recommend changing line ~197 in ShopHelper.cs :
`string str = "TownNPCMood_" + NPCID.... blah`
to be appx:
`string str = "TownNPCMood_" + (isModdedNPC(_currentNPC) ? "Mods." : "") + NPCID.... blah`

Realistically, I don't see why we can't just play with _currentNPCBeingTalkedTo and isMod checks in that method to get the correct translation key string instead of changing mod localization key template.

kittenchilly committed:
Revert localization changes.

kittenchilly committed:
Update ShopHelper.cs.patch

kittenchilly committed:
Cleanup

kittenchilly committed:
Remove ref from primaryPlayerBiome

kittenchilly committed:
More dumb cleanup

kittenchilly committed:
You know the drill

kittenchilly committed:
More ExampleMod documentation

Chicken-Bones reviewed (approved):
Looks solid to me, just that `TODO: documentation` should probably be replaced with a simple comment saying `make the guide like example person`

kittenchilly committed:
Merge branch '1.4' into 1.4_happiness

kittenchilly committed:
Fix ExampleMod

kittenchilly committed:
Merge branch '1.4' into 1.4_happiness

kittenchilly committed:
Update ShopHelper.cs.patch

Solxanich commented on ExampleMod/Content/NPCs/ExamplePerson.cs:
```diff
@@ -138,6 +139,30 @@ public class ExamplePerson : ModNPC
 
+		//Set Example Person's biome and neighbor preferences with the NPCHappiness hook. You can add happiness text and remarks with localization (See an example in ExampleMod/Localization/en-US.lang).
+		public override void NPCHappiness(int primaryPlayerBiome, ref ShopHelper shopHelperInstance, ref bool[] nearbyNPCsByType) {
+			//TODO: Make it prefer ExampleBiome when that happens. And BiomeID.
```
This TODO should be complete as #1205  has-been merged and includes all the stuff to setup for this.
Should be a simple todo to do

Solxanich committed:
Delete en-US.lang

Solxanich committed:
Merge pull request #4 from tModLoader/1.4

1.4

Solxanich committed:
Re-commit my changes

ExampleBiome for ExamplePerson
localization port

Solxanich committed:
Complete NPC Happiness PR

Mirsario commented on patches/tModLoader/Terraria/GameContent/ShopHelper.cs.patch:
```diff
@@ -32,3 +32,13 @@
  		}
  
  		private void AddHappinessReportText(string textKeyInCategory, object substitutes = null) {
++			string str;
++			if (_currentNPCBeingTalkedTo.type <= NPCID.Count)
+-			string str = "TownNPCMood_" + NPCID.Search.GetName(_currentNPCBeingTalkedTo.netID);
++				str = "TownNPCMood_" + NPCID.Search.GetName(_currentNPCBeingTalkedTo.netID);
++			else
++				str = "Mods." + _currentNPCBeingTalkedTo.ModNPC.Mod.Name + "." + _currentNPCBeingTalkedTo.ModNPC.Name + "." + "TownNPCMood";
++
```
Cache search result, use interpolated strings, using a local for modnpc would also be nice.

Solxanich replied:
Resolved exempting caching search result, not understood what to do with that.
Should be sufficient now

Mirsario commented on patches/tModLoader/Terraria/ModLoader/GlobalNPC.cs:
```diff
@@ -522,6 +523,16 @@ public virtual void DrawBehind(NPC npc, int index)
 		public virtual void SetupTravelShop(int[] shop, ref int nextSlot) {
 		}
 
+		/// <summary>
+		/// Allows you to set an NPC's biome preferences and nearby npc preferences for the NPC happiness system. Recommended to only be used with NPCs that have shops.
+		/// </summary>
+		/// <param name="npc"></param>
+		/// <param name="shopHelperInstance"></param>
+		/// <param name="primaryPlayerBiome"></param>
+		/// <param name="nearbyNPCsByType"></param>
+		public virtual void NPCHappiness(NPC npc, int primaryPlayerBiome, ref ShopHelper shopHelperInstance, ref bool[] nearbyNPCsByType) {
+		}
+
```
Document parameters. What's with the last two parameters being by ref? The last parameter could be a readonlyspan.

Solxanich replied:
Resolved

Solxanich committed:
Address Feedback from PR

Add docs, cleanup args for hook
Switch order of localization
PrimaryBiomeID so that docs make sense
Some caching and string interpolation in ShopHelper.AddHappinessReportText

Solxanich commented on ExampleMod/Localization/en-US.hjson:
```diff
@@ -62,5 +73,28 @@ Mods: {
+		#This is what displays during all of an NPC's happiness levels. Note that the localization key for these should always be "*NPCName*.TownNPCMood.*Mood*" or else it won't display properly.
+		ExamplePerson: {
+			TownNPCMood: {
+				Content: I feel pretty fine right now.
```
Resolved

Mirsario reviewed (changes requested):
A summary of my problems with the current (Vanilla) design for happiness definition:
- The if checks for `nearbyNPCsByType` and `primaryPlayerBiome` are pretty raw boilerplate. Have a look: 
![image](https://user-images.githubusercontent.com/6795251/141801389-ede6b23e-801c-4f77-a0c9-82451cf82cae.png)

- The nature of vanilla's `(Like/Dislike/Love/Hate)(NPC/Biome)` methods results in inability for mods to properly cancel out each other's effects, or do the same thing without conflicts. For example, if 3 mods make the same NPC like another same NPC, then the effect of those calls on `_currentPriceAdjustment` will be tripled, as anyone can see from looking at their code: ![image](https://user-images.githubusercontent.com/6795251/141800988-06aecc38-1793-475a-ac68-aac419672ada.png)

I would heavily prefer if we redo the vanilla system to be more declarative.

Solxanich committed:
Merge Conflicts Stage 1 - Initial Merge

Solxanich committed:
Merge branch 'tModLoader-1.4' into 1.4_happiness

Mirsario committed:
Rewrote npc/biome relationships to be declarative.

Mirsario committed:
Renamed the GlobalNPC hook too.

Mirsario committed:
Replaced NPCHappiness' constructor with a static Get.

Solxanich committed:
Fix NPC Happiness for Biomes

Mirsario commented:
Solved my above concerns with a redesign of the PR. Here's some extra TODOs:

- [ ] ModifyNPCHappiness:
    - [ ] Should contain `ref float` of price scaling.
    - [ ] nearby npcs should be a ReadOnlySpan<bool>.
    - [ ] ShopHelper likely has its remark-raising methods protected and is utterly useless. The hook should be able to modify remarks and add new ones, but I can't picture the best place for that.

kittenchilly committed:
Merge branch '1.4' into 1.4_happiness

Solxanich committed:
Action what discussed in Discord - Defer Bonus Hook to Future

Solxanich committed:
Swap to BiomeID instead of PrimaryBiomeID

De-duplication.

Solxanich commented:
PR Priority : Release. Rationale: minimize bug submissions due to happiness button on town npcs not working. 
PR target scope: Allow modders to implement localization text for npc happiness, and basic vanilla relationships.

PR scope change request: Allow custom relationships for happiness outside of vanilla presets of Like/Love/Hate. Priority: either 'Do at some point' or 'Post Release' depending on modder demand. Rationale: may break parts of current implementation, but is a feature addition with insufficient drivers relative to key Release items.
PR scope change: rejected. Rationale: Lower priority scope addition represents scope creep; defer scope to a future PR assuming #2073.

Mergeability:
Tested: Y
Peer Reviewed: Y          Dev Reviewed: Y
Outstanding Actions: N
